src/vehicle.cpp

```diff
@@ -2271,6 +2276,98 @@ void Vehicle::HandleLoading(bool mode)
 	this->IncrementImplicitOrderIndex();
 }
 
+/**
+ * Checks whether a vehicle is waiting for automatic separation (if not,
+ * it is ready to depart)
+ */
+bool Vehicle::IsWaitingForAutomaticSeparation() const {
+	Ticks now = _date * DAY_TICKS + _date_fract;
+	return this->AutomaticSeparationIsEnabled() && this->first_order_last_departure > now;
+};
+
+/**
+ * If enabled, calculates the departure time for this vehicle based on the
+ * automatic separation feature.
+ */
+void Vehicle::UpdateAutomaticSeparation()
+{
+	/* Check this feature is enabled on the vehicle's orders */
+	if (!this->AutomaticSeparationIsEnabled()) return;
+
+	/* Only perform the separation at the first manual order (saves on storage) */
+	VehicleOrderID first_manual_order = 0;
+	for (Order *o = this->GetFirstOrder(); o != nullptr && o->IsType(OT_IMPLICIT); o = o->next) {
+		++first_manual_order;
+	}
+	if (this->cur_implicit_order_index != first_manual_order) return;
+
+	/* A "last departure" >= now means we've already calculated the separation */
+	Ticks now = _date * DAY_TICKS + _date_fract;
+	if (this->first_order_last_departure >= now) return;
+
+	/* Calculate round trip time from last departure and now - automatic separation waiting time is not included */
+	if (this->first_order_last_departure > 0) {
+		this->first_order_round_trip_time = now - this->first_order_last_departure;
+	}
+
+	/* To work out the automatic separation waiting time we need to know:
+	 *   - When the last vehicle departed or will depart
+	 *   - Average time to perform the order list (as sum/count)
+	 *   - How many vehicles are currently operating the order list
+	 *   - How many vehicles are currently queuing for the first manual order
+	 */
+	Ticks last_departure = 0;
+	Ticks round_trip_sum = 0;
+	int round_trip_count = 0;
+	int vehicles = 0;
+	int vehicles_queuing = 0;
+	Vehicle *v = this->FirstShared();
+	while (v != nullptr) {
+		last_departure = max(last_departure, v->first_order_last_departure);
+		if (v->first_order_round_trip_time > 0) {
+			round_trip_sum += v->first_order_round_trip_time;
+			round_trip_count++;
+		}
+		/* A stopped vehicle is not included; it might be stopped by player or parked in a depot */
+		if (!(v->vehstatus & VS_STOPPED)) {
+			vehicles++;
+			/* Count vehicles queing for the first manual order but not currently in the station */
+			if (v != this && v->cur_speed == 0 && v->cur_implicit_order_index == first_manual_order && !v->current_order.IsType(OT_LOADING)) {
```

Comment:
What's the rationale for this heuristic?

That a vehicle is stopped somewhere does not necessarily imply that it is queuing immediately before the target station.
If a train is quite normally stopped at a red signal en-route it would have a disproportionate effect on the separation.

Reply:
This heuristic is the closest I could get to the desired information. If you can do better, please suggest how. :)

The comment below this code should explain why we need to count queuing vehicles, and, since the status is only checked when a vehicle becomes ready to depart, in practise it does not seem problematic to have a few (more/fewer) vehicles miscategorised here.

Reply:
I can see where you're coming from. Something like this probably needs play-testing on a variety of network types to see how it performs.

Given that separation is only applied at the first manual order, that order seems likely to be intended to be a terminus of some kind. The comments imply that you're thinking of a single platform type layout with a queue to get in. This would seem like an odd design choice at the player level.
Any congestion on the approach to the first manual order would be liable to make vehicles leave the terminus early. This could disproportionately impact the simple case where the order list has only two stations in it, for example.

Reply:
I suggested this elsewhere, but the ideal place to separate vehicles is a depot. I don't think it would be unreasonable to only separate at the first depot encountered by a train, with the Autoseparate button giving the user a command error if the order list does not include a depot. That would avoid the primary complaint/concern I've heard about  autoseparation, which is clogs due to vehicles attempting to space themselves out.

Reply:
Using a depot like this is a bit unintuitive if the cargo flow is bidirectional (i.e passengers, mail).
You don't really want your passengers/mail ageing in a depot waiting for a departure slot to be able to unload.
You could use an unload order, then a depot order, then a load order, but users would reasonably ask why they can't just use a terminus with enough platforms instead.

Reply:
Or let CargoDist handle it automatically, but yes, users might complain. That said, it's still the most durable way to handle waiting time.

Reply:
If we don't think the queuing applies to other types of vehicles, I am happy to condition it, but it is required for road vehicles from playtesting.